modules/packages/LinearAlgebra.chpl

```diff
@@ -927,6 +927,28 @@ proc trace(A: [?D] ?eltType) {
   return trace;
 }
 
+/* Return the Kronecker Product of matrix ``A`` and matrix ``B``.
+   If the size of A is ``x * y`` and of B is ``a * b`` then size of the resulting
+   matrix will be ``(x * a) * (y * b)`` */
+proc kron(A: [?ADom] ?eltType, B: [?BDom] eltType) {
+  if ADom.rank != 2 || BDom.rank != 2 then compilerError("Rank size not 2");
+  const (rowA, colA) = A.shape;
+  const (rowB, colB) = B.shape;
+  var C = Matrix(rowA*rowB, colA*colB, eltType=eltType);
+  forall i in 0..#rowA {
+    forall j in 0..#colA {
```

Comment:
I am on the fence about leaving this 2nd `forall` vs. changing it back to `for` loop. It could benefit performance in cases where `rowA < numPUs`. 

Does anyone from @chapel-lang/perf-team have an opinion?

Reply:
but I want to understand it, the 2nd forall loop and 1st forall loop will do the thing, parallely multiplying an integer with the matrix

Reply:
like kron(A, B) is needed, then A[i][j] * B is calculated parallely, if we use 2 forall loops, it will be fast, if we have enough number of threads, which should be equal to the size of matrix A, but now considering the complexity of algo, A would have size atmost 100 * 100, because then it will cause 10^8 ops, this may take 1s of time, so I think it will fast the things,
Where am I incorrect?

Reply:
the loop 3 and  4 is understandable as getting huge number of threads cann be a pain in optimization, but can avg 10^4 threads we get, if threads use one one model, then 10^4 LWP

Reply:
This is a good question. First, I suggest taking a look at the [`forall` section of the users guide](https://chapel-lang.org/docs/latest/users-guide/datapar/forall.html) if you haven't yet.

>  if we use 2 forall loops, it will be fast, if we have enough number of threads,

This is why I was on the fence about this. If there are more elements in `0..#rowA` than available processors, then all the available processors would be saturated with tasks from the outer-most loop, making the inner `forall` effectively a `for` loop, but with a little extra overhead from querying the available resources. The best choice depends on how small that extra overhead is and how often `numProcs > rowA`, which is hard to predict.

Fortunately, we can simplify this decision by making the outer `forall` over a 2-dimensional domain `{0..#rowA, 0..#colA}` instead of 2 nested 1-dimensional loops (see my other comment).